lib/plugin-loader.js

```diff
@@ -0,0 +1,286 @@
+/**
+ * Provides a way to load "plugins" as provided by the user.
+ *
+ * Currently supports:
+ *
+ * - Root hooks
+ * - Global fixtures (setup/teardown)
+ * @private
+ * @module plugin
+ */
+
+'use strict';
+
+const debug = require('debug')('mocha:plugin-loader');
+const {
+  createInvalidPluginDefinitionError,
+  createInvalidPluginImplementationError
+} = require('./errors');
+const {castArray} = require('./utils');
+
+/**
+ * Built-in plugin definitions.
+ */
+const MochaPlugins = [
+  /**
+   * Root hook plugin definition
+   * @type {PluginDefinition}
+   */
+  {
+    exportName: 'mochaHooks',
+    optionName: 'rootHooks',
+    validate(value) {
+      if (
+        Array.isArray(value) ||
+        (typeof value !== 'function' && typeof value !== 'object')
+      ) {
+        throw createInvalidPluginImplementationError(
+          `mochaHooks must be an object or a function returning (or fulfilling with) an object`
+        );
+      }
+    },
+    async finalize(rootHooks) {
```

Comment:
note the other two plugin definitions do _not_ have a "finalize" implementation. it's needed here (this code is not new; it came from elsewhere) because of the deeply-nested data structure and the ability to define hooks asynchronously (the user supplies an async function which returns an object containing root hooks).

global setup/teardown fixtures can be async, of course, but I don't yet see a legit use-case for allowing a function _which returns_ a fixture.